src/lib/Splint.hs

```diff
@@ -37,18 +41,66 @@ action commandLineOptions modSummary hsParsedModule = do
 type Settings = (HLint.ParseFlags, [HLint.Classify], HLint.Hint)
 
 getSettings :: [GHC.CommandLineOption] -> GHC.Hsc Settings
-getSettings commandLineOptions = GHC.liftIO $ do
-  maybeSettings <- IORef.readIORef settingsRef
-  case maybeSettings of
-    Just settings -> pure settings
-    Nothing -> do
-      settings <- HLint.argsSettings commandLineOptions
-      IORef.writeIORef settingsRef $ Just settings
-      pure settings
+getSettings commandLineOptions = do
+  remoteData <- io . stm $ do
+    settings <- Stm.readTVar settingsTVar
+    let remoteData = Map.findWithDefault NotAsked commandLineOptions settings
+    case remoteData of
+      NotAsked ->
+        Stm.modifyTVar settingsTVar $ Map.insert commandLineOptions Loading
+      _ -> pure ()
+    pure remoteData
+  case remoteData of
+    NotAsked -> io . withTMVar settingsTMVar . const $ do
```

Comment:
I think two threads could still end up here. Thanks to the `TMVar` they wouldn't both run at the same time, but they would both eventually get the settings.

Reply:
Actually no, that's not possible. The previous STM transaction starting on line 45 will update the `RemoteData` to `Loading`, so only one thread will end up in this `NotAsked` branch.